fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/service/LoanProductWritePlatformServiceJpaRepositoryImpl.java

```diff
@@ -304,6 +319,32 @@ private boolean anyChangeInCriticalFloatingRateLinkedParams(JsonCommand command,
         return charges;
     }
 
+    private List<Rate> assembleListOfProductRates(final JsonCommand command) {
+
+        final List<Rate> rates = new ArrayList<>();
+
+        if (command.parameterExists("rates")) {
+            final JsonArray ratesArray = command.arrayOfParameterNamed("rates");
+            if (ratesArray != null) {
+                for (int i = 0; i < ratesArray.size(); i++) {
+
+                    final JsonObject jsonObject = ratesArray.get(i).getAsJsonObject();
+                    if (jsonObject.has("id")) {
+                        final Long id = jsonObject.get("id").getAsLong();
+
+                        final Rate rate = this.rateRepository.findOne(id);
```

Comment:
Instead of repeated calls to DB, can you try using single query with in clause and validate